File: src/main/java/io/github/daviddenton/metrique/metrics/Metrics.java
package io.github.daviddenton.metrique.metrics;

import java.util.List;

public class Metrics {

    private final List<String> rootName;
    private final MetricsClient client;
    public final MetricName name;

    protected Metrics(MetricsClient client, List<String> rootName) {
        this.rootName = rootName;
        this.client = client;
        this.name = new MetricName(rootName);
    }

    public Metrics child(String... newParts) {
        return new Metrics(client, name.getCollect(newParts, rootName));
    }

    public Metric metric(String... newParts) {
        MetricName finalName = child(newParts).name;

        return new Metric() {
            public void count(Long value) {
                client.count(finalName, value);
            }

            public void increment() {
                client.increment(finalName);
            }

            public void decrement() {
                client.decrement(finalName);
            }

            public void gauge(Long value) {
                client.gauge(finalName, value);
            }

            public void time(Long value) {
                client.time(finalName, value);
            }
        };
    }
}

File: src/main/java/io/github/daviddenton/metrique/metrics/Metrics.scala
package io.github.daviddenton.metrique.metrics

import springagle.cloudfoundry.CfAppInstanceIndex
import springagle.util.{Environment, SystemName}

class Metrics private(parts: List[String], client: MetricsClient) {

  def this(systemName: SystemName, env: Environment, appInstanceIndex: CfAppInstanceIndex, client: MetricsClient) =
    this(List[String]("services", systemName.value, env.name, appInstanceIndex.index.toString).reverse, client)

  private val name = MetricName(parts.reverse.mkString("."))

  def child(newParts: String*): Metrics = new Metrics(newParts.reverse.toList ++ parts, client)

  def metric(newParts: String*): Metric = {
    val finalName = child(newParts: _*).name

    new Metric {
      override def count(value: Long): Unit = client.count(finalName, value)

      override def increment(): Unit = client.increment(finalName)

      override def decrement(): Unit = client.decrement(finalName)

      override def gauge(value: Long): Unit = client.gauge(finalName, value)

      override def time(value: Long): Unit = client.time(finalName, value)
    }
  }
}
